### PR TITLE
Adding implementation for instance_state and view_state [12758]

### PR DESCRIPTION
**Pull request by MiguelCompany**
Implementing figure 2.11 (Statechart of the instance_state and view_state for a single instance) of the DDS standard.

### CODE REVIEW

File: include/fastrtps/utils/collections/ResourceLimitedVector.hpp [hunk 1]
```diff
@@ -119,6 +119,46 @@ class ResourceLimitedVector
...
+    iterator insert(
+            const_iterator pos,
+            value_type&& value)
+    {
+        if (capacity() >= configuration_.maximum)
+        {
+            return end();
+        }
+
+        return collection_.insert(pos, value);
```
Comment (MiguelBarro) [inline include/fastrtps/utils/collections/ResourceLimitedVector.hpp:159]: value is now a left value.
```suggestion
        return collection_.insert(std::move(pos), value);
```

File: src/cpp/fastdds/subscriber/history/DataReaderInstance.hpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+namespace eprosima {
+namespace fastdds {
+namespace dds {
+namespace detail {
+
+/// Book-keeping information for an instance
+struct DataReaderInstance
+{
+    using ChangeCollection = eprosima::fastrtps::ResourceLimitedVector<DataReaderCacheChange, std::true_type>;
+    using WriterCollection = std::map<fastrtps::rtps::GUID_t, uint32_t>;
```
Comment (MiguelBarro) [inline src/cpp/fastdds/subscriber/history/DataReaderInstance.hpp:42]: The use of a std::map breaks the zero dynamic allocation policy.

File: include/fastdds/rtps/history/ReaderHistory.h [hunk 3]
```diff
@@ -132,6 +138,38 @@ class ReaderHistory : public History
...
 
+    template<typename Pred>
+    inline bool remove_changes_with_pred(
+            Pred pred)
+    {
+        assert(nullptr != mp_reader);
+        assert(nullptr != mp_mutex);
+
+        // Lock scope
+        std::vector<CacheChange_t*> changes_to_remove;
```
Comment (MiguelBarro) [inline include/fastdds/rtps/history/ReaderHistory.h:149]: The use of a std::vector here breaks zero allocation policy.

File: include/fastdds/rtps/history/ReaderHistory.h [hunk 4]
```diff
@@ -117,6 +117,18 @@ class ReaderHistory : public History
             CacheChange_t** min_change,
             const GUID_t& writerGuid);
 
+    RTPS_DllAPI virtual bool writer_unmatched(
```
Comment (MiguelBarro) [inline include/fastdds/rtps/history/ReaderHistory.h:120]: Pending doxygen.

**Review by MiguelBarro — CHANGES_REQUESTED** [review]
+ Some tests trigger assertions when the WriterProxies are notified of sample removal (history >> reader >> proxies).
+ The DataReaderTests.check_key_history_wholesomeness_on_unmatch must be adapted to the new standard-compliant behaviour (remove only non-notified samples on writer unmatch).
+ New tests must be added to check the generation and state member of the retrieved collections' SampleInfos.

File: include/fastrtps/utils/collections/ResourceLimitedVector.hpp (hunk 1 above)
Comment (MiguelCompany) [reply include/fastrtps/utils/collections/ResourceLimitedVector.hpp:159]: I guess you mean `collection_.insert(pos, std::move(value));`

**Comment by MiguelCompany** [comment]
> * New tests must be added to check the generation and state member of the retrieved collections' SampleInfos.

We already added them on #2193, which is already merged.

File: include/fastrtps/utils/collections/ResourceLimitedVector.hpp (hunk 1 above)
Comment (MiguelBarro) [reply include/fastrtps/utils/collections/ResourceLimitedVector.hpp:159]: exactly 😅

File: src/cpp/fastdds/subscriber/history/DataReaderInstance.hpp (hunk 2 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/history/DataReaderInstance.hpp]: Changed into a ResourceLimitedVector of pairs

File: include/fastdds/rtps/history/ReaderHistory.h (hunk 3 above)
Comment (MiguelCompany) [reply include/fastdds/rtps/history/ReaderHistory.h]: Refactored into using std::remove_if

File: include/fastdds/rtps/history/ReaderHistory.h (hunk 4 above)
Comment (MiguelCompany) [reply include/fastdds/rtps/history/ReaderHistory.h]: done

**Comment by MiguelCompany** [comment]
@richiprosima Please test aarch64

**Comment by MiguelBarro** [comment]
@richiprosima Please test windows

**Comment by MiguelCompany** [comment]
@richiprosima Please test aarch64

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

File: src/cpp/fastdds/subscriber/history/DataReaderHistory.hpp [hunk 5]
```diff
@@ -161,14 +207,25 @@ class DataReaderHistory : public eprosima::fastrtps::rtps::ReaderHistory
...
+    void writer_not_alive(
+            const fastrtps::rtps::GUID_t& writer_guid);
+
+    void check_and_remove_instance(
+            instance_info& instance_info);
+
 private:
 
-    using t_m_Inst_Caches = std::map<InstanceHandle_t, fastrtps::KeyedChanges>;
+    using InstanceCollection = std::map<InstanceHandle_t, DataReaderInstance>;
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/history/DataReaderHistory.hpp:221]: The commitment was with topics without keys. That is why we are creating the 'unknown' instance in the constructor when the topic has no keys, to have it pre-allocated.

For a future minor release, we could consider having an `initial_instances` configuration parameter on the resource limits QoS.

File: src/cpp/fastdds/subscriber/history/DataReaderHistory.hpp [hunk 6]
```diff
@@ -34,12 +34,16 @@
 #include <fastdds/dds/topic/TypeSupport.hpp>
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/history/DataReaderHistory.hpp:33]: Done in 134afc6

File: src/cpp/fastdds/subscriber/history/DataReaderHistory.hpp [hunk 7]
```diff
@@ -161,14 +207,25 @@ class DataReaderHistory : public eprosima::fastrtps::rtps::ReaderHistory
             const InstanceHandle_t& handle,
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/history/DataReaderHistory.hpp:205]: Done in ed04235

File: src/cpp/fastdds/subscriber/history/DataReaderHistory.hpp [hunk 8]
```diff
@@ -216,29 +267,34 @@ class DataReaderHistory : public eprosima::fastrtps::rtps::ReaderHistory
...
             CacheChange_t* change,
-            size_t unknown_missing_changes_up_to);
+            DataReaderInstance& instance);
 
-    bool received_change_keep_last_with_key(
+    bool completed_change_keep_last(
             CacheChange_t* change,
-            size_t unknown_missing_changes_up_to);
+            DataReaderInstance& instance);
     ///@}
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/history/DataReaderHistory.hpp:293]: Done in ed04235

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM

**Comment by MiguelCompany** [comment]
Leaving CI for the PR into master.

**Pull request merged**